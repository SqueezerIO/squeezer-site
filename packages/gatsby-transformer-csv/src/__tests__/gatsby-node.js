const Promise = require(`bluebird`)
const _ = require(`lodash`)
const json2csv = require(`json2csv`)

const { onCreateNode } = require(`../gatsby-node`)

describe(`Process  nodes correctly`, () => {
  const node = {
    id: `whatever`,
    parent: `SOURCE`,
    children: [],
    extension: `csv`,
    internal: {
      contentDigest: `whatever`,
      mediaType: `text/csv`,
    },
    name: `test`,
  }

  // Make some fake functions its expecting.
  const loadNodeContent = node => Promise.resolve(node.content)

  it(`correctly creates nodes from JSON which is an array of objects`, async () => {
    const fields = [`blue`, `funny`]
    const data = [{ blue: true, funny: `yup` }, { blue: false, funny: `nope` }]
    const csv = json2csv({ data: data, fields: fields })
    node.content = csv

    const createNode = jest.fn()
    const createParentChildLink = jest.fn()
    const boundActionCreators = { createNode, createParentChildLink }

    await onCreateNode({
      node,
      loadNodeContent,
      boundActionCreators,
    }).then(() => {
      expect(createNode.mock.calls).toMatchSnapshot()
      expect(createParentChildLink.mock.calls).toMatchSnapshot()
      expect(createNode).toHaveBeenCalledTimes(2)
      expect(createParentChildLink).toHaveBeenCalledTimes(2)
    })
  })

  it(`correctly handles the options object that is passed to it`, async () => {
    node.content = `blue,funny\ntrue,yup\nfalse,nope`

    const createNode = jest.fn()
    const createParentChildLink = jest.fn()
    const boundActionCreators = { createNode, createParentChildLink }

    await onCreateNode(
      {
        node,
        loadNodeContent,
        boundActionCreators,
      },
      { noheader: true }
    ).then(() => {
      expect(createNode.mock.calls).toMatchSnapshot()
      expect(createParentChildLink.mock.calls).toMatchSnapshot()
      expect(createNode).toHaveBeenCalledTimes(3)
      expect(createParentChildLink).toHaveBeenCalledTimes(3)
    })
  })

  it(`If the object has an id, it uses that as the id instead of the auto-generated one`, async () => {
    const fields = [`id`, `blue`, `funny`]
    const data = [
      { id: `foo`, blue: true, funny: `yup` },
      { blue: false, funny: `nope` },
    ]
    const csv = json2csv({ data: data, fields: fields })
    node.content = csv

    const createNode = jest.fn()
    const createParentChildLink = jest.fn()
    const boundActionCreators = { createNode, createParentChildLink }

    await onCreateNode({
      node,
      loadNodeContent,
      boundActionCreators,
    }).then(() => {
      expect(createNode.mock.calls[0][0].id).toEqual(`foo`)
    })
  })

  it(`the different objects shouldn't get the same ID even if they have the same content`, async () => {
    const fields = [`id`, `blue`, `funny`]
    const data = [
      { id: `foo`, blue: true, funny: `yup` },
      { blue: false, funny: `nope` },
      { blue: false, funny: `nope` },
      { green: false, funny: `nope` },
    ]
    const csv = json2csv({ data: data, fields: fields })
    node.content = csv

    const createNode = jest.fn()
    const createParentChildLink = jest.fn()
    const boundActionCreators = { createNode, createParentChildLink }

    await onCreateNode({
      node,
      loadNodeContent,
      boundActionCreators,
    }).then(() => {
      const ids = createNode.mock.calls.map(object => object[0].id)
      // Test that they're unique
      expect(_.uniq(ids).length).toEqual(4)
    })
  })
})
