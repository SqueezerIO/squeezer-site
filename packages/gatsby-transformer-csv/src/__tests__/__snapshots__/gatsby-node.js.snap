// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Process  nodes correctly correctly creates nodes from JSON which is an array of objects 1`] = `
Array [
  Array [
    Object {
      "blue": "true",
      "children": Array [],
      "funny": "yup",
      "id": "whatever [0] >>> CSV",
      "internal": Object {
        "contentDigest": "5005aee6b2557974cae0aabc712e125a",
        "type": "TestCsv",
      },
      "parent": "whatever",
    },
  ],
  Array [
    Object {
      "blue": "false",
      "children": Array [],
      "funny": "nope",
      "id": "whatever [1] >>> CSV",
      "internal": Object {
        "contentDigest": "0af87da7e572c2d6d8493bcd642c2c78",
        "type": "TestCsv",
      },
      "parent": "whatever",
    },
  ],
]
`;

exports[`Process  nodes correctly correctly creates nodes from JSON which is an array of objects 2`] = `
Array [
  Array [
    Object {
      "child": Object {
        "blue": "true",
        "children": Array [],
        "funny": "yup",
        "id": "whatever [0] >>> CSV",
        "internal": Object {
          "contentDigest": "5005aee6b2557974cae0aabc712e125a",
          "type": "TestCsv",
        },
        "parent": "whatever",
      },
      "parent": Object {
        "children": Array [],
        "content": "\\"blue\\",\\"funny\\"
true,\\"yup\\"
false,\\"nope\\"",
        "extension": "csv",
        "id": "whatever",
        "internal": Object {
          "contentDigest": "whatever",
          "mediaType": "text/csv",
        },
        "name": "test",
        "parent": "SOURCE",
      },
    },
  ],
  Array [
    Object {
      "child": Object {
        "blue": "false",
        "children": Array [],
        "funny": "nope",
        "id": "whatever [1] >>> CSV",
        "internal": Object {
          "contentDigest": "0af87da7e572c2d6d8493bcd642c2c78",
          "type": "TestCsv",
        },
        "parent": "whatever",
      },
      "parent": Object {
        "children": Array [],
        "content": "\\"blue\\",\\"funny\\"
true,\\"yup\\"
false,\\"nope\\"",
        "extension": "csv",
        "id": "whatever",
        "internal": Object {
          "contentDigest": "whatever",
          "mediaType": "text/csv",
        },
        "name": "test",
        "parent": "SOURCE",
      },
    },
  ],
]
`;

exports[`Process  nodes correctly correctly handles the options object that is passed to it 1`] = `
Array [
  Array [
    Object {
      "children": Array [],
      "field1": "blue",
      "field2": "funny",
      "id": "whatever [0] >>> CSV",
      "internal": Object {
        "contentDigest": "49605f5e32a2250d7ca740c5d2e79784",
        "type": "TestCsv",
      },
      "parent": "whatever",
    },
  ],
  Array [
    Object {
      "children": Array [],
      "field1": "true",
      "field2": "yup",
      "id": "whatever [1] >>> CSV",
      "internal": Object {
        "contentDigest": "2bf0bc0e8539ef5f6bbcf67c8b733233",
        "type": "TestCsv",
      },
      "parent": "whatever",
    },
  ],
  Array [
    Object {
      "children": Array [],
      "field1": "false",
      "field2": "nope",
      "id": "whatever [2] >>> CSV",
      "internal": Object {
        "contentDigest": "f5f617ce4243d639de924543dec5600e",
        "type": "TestCsv",
      },
      "parent": "whatever",
    },
  ],
]
`;

exports[`Process  nodes correctly correctly handles the options object that is passed to it 2`] = `
Array [
  Array [
    Object {
      "child": Object {
        "children": Array [],
        "field1": "blue",
        "field2": "funny",
        "id": "whatever [0] >>> CSV",
        "internal": Object {
          "contentDigest": "49605f5e32a2250d7ca740c5d2e79784",
          "type": "TestCsv",
        },
        "parent": "whatever",
      },
      "parent": Object {
        "children": Array [],
        "content": "blue,funny
true,yup
false,nope",
        "extension": "csv",
        "id": "whatever",
        "internal": Object {
          "contentDigest": "whatever",
          "mediaType": "text/csv",
        },
        "name": "test",
        "parent": "SOURCE",
      },
    },
  ],
  Array [
    Object {
      "child": Object {
        "children": Array [],
        "field1": "true",
        "field2": "yup",
        "id": "whatever [1] >>> CSV",
        "internal": Object {
          "contentDigest": "2bf0bc0e8539ef5f6bbcf67c8b733233",
          "type": "TestCsv",
        },
        "parent": "whatever",
      },
      "parent": Object {
        "children": Array [],
        "content": "blue,funny
true,yup
false,nope",
        "extension": "csv",
        "id": "whatever",
        "internal": Object {
          "contentDigest": "whatever",
          "mediaType": "text/csv",
        },
        "name": "test",
        "parent": "SOURCE",
      },
    },
  ],
  Array [
    Object {
      "child": Object {
        "children": Array [],
        "field1": "false",
        "field2": "nope",
        "id": "whatever [2] >>> CSV",
        "internal": Object {
          "contentDigest": "f5f617ce4243d639de924543dec5600e",
          "type": "TestCsv",
        },
        "parent": "whatever",
      },
      "parent": Object {
        "children": Array [],
        "content": "blue,funny
true,yup
false,nope",
        "extension": "csv",
        "id": "whatever",
        "internal": Object {
          "contentDigest": "whatever",
          "mediaType": "text/csv",
        },
        "name": "test",
        "parent": "SOURCE",
      },
    },
  ],
]
`;
